stream: Add subscriber_count field.

abdelrahman725 opened the pull request:
Fixes #34246

Prerequisite to #34244

1.  Add `subscriber_count` field to `Stream` to keep track of number of non-deactivated users subscribed to the channel.

2. New functions in "zerver/lib/stream_subscription.py" for handling different cases of updating `subscriber_count` field, letting all updates happen only at the DB-level, to avoid some race conditions.

3.  Data migration file, a single SQL query to compute the delta per stream and update the stream table with that computed value, the query is batched by one batch per realm.

abdelrahman725 commented on zerver/lib/stream_subscription.py:
```diff
@@ -313,3 +313,19 @@ def get_subscriptions_for_send_message(
     return query
+
+
+def update_stream_subscriber_count(
+    stream_to_subscribers: dict[Stream, set[UserProfile]], count_sign: Literal[1, -1]
+) -> None:
+    """increment/decrement subscriber_count for non-deactivated stream subscribers.
+
+    count_sign -> 1=increment, -1=decrement
+    """
+
+    for stream, subscribers in stream_to_subscribers.items():
+        stream.subscriber_count += (
+            len([user.id for user in subscribers if user.is_active]) * count_sign
+        )
+
+    Stream.objects.bulk_update(stream_to_subscribers.keys(), ["subscriber_count"])
```
For further optimization, we can add a new parameter `active_subscribers` which callers can pass as `True` if they are sure that all passed subscribers are already active, then we won't need to check for `user.is_active` in the loop

timabbott replied:
I would also look hard at whether we can do a database query that asks the database to do the deltas; that might avoid certain classes of races?

timabbott commented on zerver/migrations/0698_stream_subscriber_count.py:
```diff
@@ -0,0 +1,17 @@
+        migrations.AddField(
+            model_name="stream",
+            name="subscriber_count",
+            field=models.PositiveIntegerField(default=0),
```
you'll want to use `db_default=0` to avoid this migration needing to rewrite every `Stream` row immediately.

timabbott commented on zerver/lib/stream_subscription.py:
```diff
@@ -313,3 +313,19 @@ def get_subscriptions_for_send_message(
+) -> None:
+    """increment/decrement subscriber_count for non-deactivated stream subscribers.
+
+    count_sign -> 1=increment, -1=decrement
```
`direction` would be a better name for `count_sign`.

timabbott commented on zerver/actions/streams.py:
```diff
@@ -787,9 +788,11 @@ def bulk_add_subscriptions(
     for sub_info in subs_to_add + subs_to_activate:
         altered_user_dict[sub_info.stream.id].add(sub_info.user.id)
         altered_streams_dict[sub_info.user].add(sub_info.stream.id)
+        stream_to_subscribers[sub_info.stream].add(sub_info.user)
```
I think it might be better to only add the user ID to the list if it `is_active` here -- saves putting large structures into a new set.

timabbott commented on zerver/actions/streams.py:
```diff
@@ -787,9 +788,11 @@ def bulk_add_subscriptions(
     altered_user_dict: dict[int, set[int]] = defaultdict(set)
     altered_guests: set[int] = set()
     altered_streams_dict: dict[UserProfile, set[int]] = defaultdict(set)
+    stream_to_subscribers: dict[Stream, set[UserProfile]] = defaultdict(set)
```
`subscriber_count_changes` seems like a better name.

abdelrahman725 replied:
I don't agree, but ok

timabbott replied:
That doesn't at all communicate that it's a map of deltas?

timabbott commented on zerver/actions/streams.py:
```diff
@@ -1054,13 +1059,20 @@ def get_non_subscribed_subs() -> list[tuple[UserProfile, Stream]]:
+
         # Log subscription activities in RealmAuditLog
         event_time = timezone_now()
         event_last_message_id = get_last_message_id()
```
These actions changes seem incomplete; this number should also change when users are created, deactivated, or reactivated.

abdelrahman725 commented:
# updates

1. Addressed review comments
2. All updates  to `subscriber_count` are handled directly db, to prevent race conditions.
3. Add a pre-commit, passing `update_fields` to all `stream.save()`, To be explicit in what fields we update and to prevent overwriting of `subscriber_count`  with the in-memory default value of 0, since we ONLY update `subscriber_count` directly in db.
4. increment/decrement `subscriber_count` when reactivating/deactivating a user.
5. Add a data migration file, to set  `subscriber_count` value for existing streams.

timabbott commented on zerver/actions/streams.py:
```diff
@@ -1683,7 +1683,7 @@ def do_change_stream_group_based_setting(
 
     setattr(stream, setting_name, user_group)
-    stream.save()
+    stream.save(update_fields=[setting_name, "name"])
```
Why is `name` on this list? `stream.name` was not modified.

abdelrahman725 replied:
It was modified in [test_rename_stream](https://github.com/zulip/zulip/blob/main/zerver/tests/test_subs.py#L2169), I tested that by printing the stream.name --fetched from db-- before and after `stream.save()`

timabbott replied:
I don't think `name` is a field that `do_change_stream_group_based_setting` should be handling? And I don't see a test there that passes `name` as the `setting_name` parameter. If there is one, the test is bugged, and should call the function for changing a channel name, not `do_change_stream_group_based_setting`.

abdelrahman725 replied:
`do_change_stream_group_based_setting` didn't change the name, it got changed in other code, There is a difference in names in my above test only because of  `stream.save()`.

timabbott commented on zerver/tests/test_user_topics.py:
```diff
@@ -36,7 +36,7 @@ def test_get_deactivated_muted_topic(self) -> None:
 
         stream.deactivated = True
-        stream.save()
+        stream.save(update_fields=["deactivated"])
```
Should this be using `do_deactivate_stream` instead? Usually we aim to use real code like that in tests.

timabbott replied:
In general, we've been trying to clean up tests to use action functions for creating state, so you could consider replacing the test parts of this commit with a prep PR that does that -- I think it's just the `date_created` ones that would still do hacky edits.

abdelrahman725 replied:
ok

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -2127,6 +2127,7 @@ def get_notified_user_ids() -> set[int]:
                 )
             self.assert_json_success(result)
             event = events[0]["event"]
+            # here !
```
This looks like stale debug code?

abdelrahman725 replied:
yes, this was in a previous version

timabbott commented on zerver/lib/stream_subscription.py:
```diff
@@ -313,3 +315,94 @@ def get_subscriptions_for_send_message(
+
+    # construct a string of tuples (stream_id, n_subscribers)
+    # to be used as the columns of the temporary table delta_table.
+    stream_delta_values = sql.SQL(", ").join(
```
Surely there's a more idiomatic way to make an SQL array of values than this.

abdelrahman725 replied:
done

timabbott commented on zerver/lib/stream_subscription.py:
```diff
@@ -313,3 +315,94 @@ def get_subscriptions_for_send_message(
     return query
+
+
+def update_user_streams_subscriber_count(activate: bool, user_profile: UserProfile) -> None:
```
I think it might be most readable to just have two small functions, `update_subscriber_counts_for_new_user(user_profile)` and `update_subscriber_counts_for_deactivated_user(user_profile)`

I'm also not sure why you filter on `is_user_active` -- that deserves a comment.

abdelrahman725 replied:
> I think it might be most readable to just have two small functions, `update_subscriber_counts_for_new_user(user_profile)` and `update_subscriber_counts_for_deactivated_user(user_profile)`

For creating a new user, there is `update_streams_subscriber_count` which is called in `bulk_add_subscriptions`, this leaves `update_user_streams_subscriber_count` more suitable for reactivating/deactivating a user.



> I'm also not sure why you filter on `is_user_active` -- that deserves a comment.

While it's not wrong, yes it's not necessary since it's already guaranteed that all user stream subscriptions will be either with `is_user_active=True` or `False`, so no need to filer by that field.

timabbott replied:
Yeah, I think it's probably clearer to not filter given that we already know that state, one way or another, in the caller.

timabbott commented on zerver/migrations/0699_stream_subscriber_count_data_migration.py:
```diff
@@ -0,0 +1,57 @@
+
+    stream_subsriptions = Subscription.objects.filter(
+        recipient__type=2,
+        recipient__type_id__in=Stream.objects.values("id"),
```
What does this line do -- does it help the query plan somehow? Logically, it should be implied by the `type=2` part.

abdelrahman725 replied:
Yes, my bad

abdelrahman725 replied:
What we do need to filter by here is  both `active=True` and `is_user_active=True`, because we want only `active` subscriptions whose members are also active i.e. `is_user_active` = True.

timabbott replied:
Correct.

abdelrahman725 commented on analytics/management/commands/populate_analytics_db.py:
```diff
@@ -117,10 +118,10 @@ def handle(self, *args: Any, **options: Any) -> None:
         stream.save(update_fields=["recipient"])
 
         # Subscribe shylock to the stream to avoid invariant failures.
-        Subscription.objects.create(
```
if you mean `Subscription.objects.create()`, then no, other codes uses it for creating other subscription types.
That's why the new function is deliberately called `create_stream_subscription` i.e creates subscription for a stream

timabbott replied:
Yes, that's what I mean.

abdelrahman725 commented on zerver/lib/stream_subscription.py:
```diff
@@ -313,3 +315,94 @@ def get_subscriptions_for_send_message(
+        color=color,
+    )
+
+    if recipient.type != Recipient.STREAM or not user_profile.is_active:
```
`if recipient.type != Recipient.STREAM` is not necessary, since this function is explicilty named (create_stream_subscription) and called only for stream subscription

timabbott replied:
Doing an `assert` is probably helpful, then,

abdelrahman725 commented:
# updates

1. Addressed review comments
2. Optimize the data migration script, execute only a single SQL query, moving the whole compuation overhead (computing the delta per stream and update the table with the computed value) load from python (slower) to the db.
3. Batch the data migration SQL, one batch per realm.

abdelrahman725 commented on zerver/tests/test_signup.py:
```diff
@@ -4392,8 +4395,15 @@ def test_deactivate_user(self) -> None:
+        expected_stream_deltas = {
+            stream_id: count - 1 for stream_id, count in stream_deltas.items()
+        }
+        self.assert_stream_subscriber_count(expected_stream_deltas)
```
I'm ok with your proposed design, but How is that related to end-to-end ? I believe my approach is still end-to-end, as I'm not testing the functions in `stream_subscription.py` alone.

We would NOT need a parameter for the sign, just a `difference` which is positive/negative.

alya commented:
@timabbott Are you reviewing this one directly?

timabbott commented on zerver/migrations/0699_stream_subscriber_count_data_migration.py:
```diff
@@ -0,0 +1,53 @@
+
+    cursor = connection.cursor()
+    for realm in Realm.objects.all():
+        cursor.execute(query, {"realm_id": realm.id})
```
You can't do this all in a single transaction; set `atomic=False` on the Migration below, like 95% of our migrations do.

I recommend reading the last ~20 migrations and improving this one based on what you see there.

abdelrahman725 replied:
> You can't do this all in a single transaction
Ok, but why ? It's always good to know the "why" to learn

timabbott commented on zerver/migrations/0699_stream_subscriber_count_data_migration.py:
```diff
@@ -0,0 +1,53 @@
+    # The whole query is then executed by one batch per realm.
+    query = sql.SQL(
+        """UPDATE {stream_table} AS stream
+            SET subscriber_count = delta_table.n_subscripers
```
`n_subscripers` is not spelled correctly.

timabbott commented on zerver/models/streams.py:
```diff
@@ -239,6 +244,7 @@ def is_history_public_to_subscribers(self) -> bool:
         "is_web_public",
         "message_retention_days",
         "name",
+        # "subscriber_count",
```
Why is this commented out?

abdelrahman725 replied:
Will be included in the next changes, when we expose this field to be consumed by client via API

timabbott replied:
Yeah, that should be in a future commit, but best to not add comments like this in a PR for review.

timabbott commented on zerver/migrations/0699_stream_subscriber_count_data_migration.py:
```diff
@@ -0,0 +1,53 @@
+                    AND subscription.active = True
+                    AND subscription.is_user_active = True
+                GROUP BY stream_id
+            ) AS delta_table
```
`delta_table` seems like a bad name; if you're going to copy/paste, be sure to actually reconsider every line of code and variable name afterwards. Otherwise, you're better off writing the code afresh.

abdelrahman725 replied:
ok, changed to "stream_subscribers_table"

timabbott commented on zerver/lib/test_helpers.py:
```diff
@@ -808,3 +816,18 @@ def dns_txt_answer(name_str: str, txt: str) -> dns.resolver.Answer:
+
+
+def fetch_other_streams_subscriber_count(stream_ids: set[int]) -> dict[int, int]:
+    return build_streams_subscriber_count(streams=Stream.objects.exclude(id__in=stream_ids))
```
I think these might be better if defined on `ZulipTestCase` so you don't need to import the functions by name just to use these test helpers.

abdelrahman725 replied:
done

timabbott commented on zerver/lib/stream_subscription.py:
```diff
@@ -76,6 +79,12 @@ def get_stream_subscriptions_for_user(user_profile: UserProfile) -> QuerySet[Sub
+def get_user_subscribed_streams(user_profile: UserProfile) -> QuerySet[Stream]:
+    return Stream.objects.filter(
+        recipient_id__in=get_stream_subscriptions_for_user(user_profile).values("recipient_id")
+    )
```
I think this would do better using `get_subscribed_stream_recipient_ids_for_user`

abdelrahman725 replied:
done

timabbott commented on zerver/lib/stream_subscription.py:
```diff
@@ -313,3 +322,87 @@ def get_subscriptions_for_send_message(
     return query
+
+
+def update_user_streams_subscriber_count(
```
`update_all_subscriber_counts_for_user` would be a better name.

timabbott commented on zerver/lib/stream_subscription.py:
```diff
@@ -313,3 +322,87 @@ def get_subscriptions_for_send_message(
+    )
+
+
+def update_streams_subscriber_count(
```
`bulk_update_subscriber_counts` seems like a better name here.

abdelrahman725 replied:
done

timabbott commented on zerver/lib/stream_subscription.py:
```diff
@@ -313,3 +322,87 @@ def get_subscriptions_for_send_message(
+
+
+@transaction.atomic(savepoint=False)
+def bulk_create_stream_subscriptions(  # nocoverage
```
This nocoverage also looks wrong.

abdelrahman725 replied:
Yes

abdelrahman725 replied:
Ok, I've add a test case for this, not directly calling the function, but instead ensuring it did it work.

But it's still missing from the test coverage report (i.e. false alarm) so we should keep the nocoverage.

abdelrahman725 commented on zerver/lib/test_classes.py:
```diff
@@ -1376,6 +1376,31 @@ def assert_message_stream_name(self, message: Message, stream_name: str) -> None
+    def assert_stream_subscriber_count(
+        self,
+        deltas_before: dict[int, int],
+        deltas_after: dict[int, int],
```
Yes, by naming conventions, they are not deltas, the difference between them is the delta.

abdelrahman725 commented on zerver/lib/stream_subscription.py:
```diff
@@ -313,3 +322,87 @@ def get_subscriptions_for_send_message(
+
+
+def update_user_streams_subscriber_count(
+    direction: Literal[1, -1], user_profile: UserProfile
```
Better ?

> increment/decrement n_subscribers by 1 for user's subscribed streams, when reactivating/deactivating user.

timabbott commented on zerver/lib/stream_subscription.py:
```diff
@@ -313,3 +322,87 @@ def get_subscriptions_for_send_message(
+        color=color,
+    )
+
+    if user_profile.is_active:  # nocoverage
```
Let's do an assert instead. I don't think there's a reason to call this with a deactivated user given how it works.

abdelrahman725 commented:
# updates
Addressed review comments 
# Upcoming updates 
Expose `subscriber_count` to be consumable to client via API.

abdelrahman725 commented on zerver/migrations/0701_stream_subscriber_count_data_migration.py:
```diff
@@ -0,0 +1,56 @@
+    # The whole query is then executed by one batch per realm.
+    query = sql.SQL(
+        """UPDATE {stream_table} AS stream
+            SET subscriber_count = stream_subscribers_table.n_subscribers
```
# The logic of this line (18) is critical
Since the batches are run in multiple transactions, the computed value `n_subscribers` should always "override" the stored value, in cases of errors and retries.

This is Aopposed to the one in `bulk_update_subscriber_counts` which "increments" the field by the computed delta.

timabbott replied:
Yes, the migration should be just setting the value equal to the current count, not doing anything with deltas.

timabbott commented on zerver/migrations/0704_stream_subscriber_count_data_migration.py:
```diff
@@ -0,0 +1,55 @@
+
+    for realm in Realm.objects.all():
+        with connection.cursor() as cursor, transaction.atomic(durable=True):
+            cursor.execute(query, {"realm_id": realm.id})
```
I did a bit of reformatting/rewriting on the comments and variable names here.

``` diff
diff --git a/zerver/lib/stream_subscription.py b/zerver/lib/stream_subscription.py
index 023e83110b..728aceabe1 100644
--- a/zerver/lib/stream_subscription.py
+++ b/zerver/lib/stream_subscription.py
@@ -348,17 +348,19 @@ def bulk_update_subscriber_counts(
     if len(streams) == 0:
         return
 
-    # list of tuples (stream_id, n_subscribers -- to be decremented/incremented)
-    # used as the columns of the temporary table delta_table.
+    # list of tuples (stream_id, delta_subscribers) used as the
+    # columns of the temporary table delta_table.
     stream_delta_values = [
         (stream_id, len(subscribers) * direction) for stream_id, subscribers in streams.items()
     ]
 
     # The goal here is to update subscriber_count in a bulk efficient way,
     # letting the database handle the deltas to avoid some race conditions.
-    # But unlike update_all_subscriber_counts_for_user which uses F() for a single delta value, we can't
-    # use F() to apply different deltas per row in a single update using ORM,
-    # so we use a raw SQL.
+    #
+    # But unlike update_all_subscriber_counts_for_user which uses F()
+    # for a single delta value, we can't use F() to apply different
+    # deltas per row in a single update using ORM, so we use a raw
+    # SQL query.
     query = sql.SQL(
         """UPDATE {stream_table}
             SET subscriber_count = {stream_table}.subscriber_count + delta_table.delta
@@ -379,8 +381,9 @@ def create_stream_subscription(
     color: str = Subscription.DEFAULT_STREAM_COLOR,
 ) -> None:
     """
-    Creates a single stream Subscription object, incrementing stream.subscriber_count by 1
-    if user is active, in the same transaction.
+    Creates a single stream Subscription object, incrementing
+    stream.subscriber_count by 1 if user is active, in the same
+    transaction.
     """
 
     # We only create a stream subscription in this function
@@ -402,7 +405,10 @@ def bulk_create_stream_subscriptions(  # nocoverage
     subs: list[Subscription], streams: dict[int, set[int]]
 ) -> None:
     """
-    Bulk create subscripions for streams, incrementing stream.subscriber_count in the same transaction.
+    Bulk create subscripions for streams, incrementing
+    stream.subscriber_count in the same transaction.
+
+    Currently only used in populate_db.
     """
     Subscription.objects.bulk_create(subs)
     bulk_update_subscriber_counts(direction=1, streams=streams)
diff --git a/zerver/migrations/0705_stream_subscriber_count_data_migration.py b/zerver/migrations/0705_stream_subscriber_count_data_migration.py
index ada48309fb..0938bc2324 100644
--- a/zerver/migrations/0705_stream_subscriber_count_data_migration.py
+++ b/zerver/migrations/0705_stream_subscriber_count_data_migration.py
@@ -10,15 +10,16 @@ def set_stream_subscribe_count(apps: StateApps, schema_editor: BaseDatabaseSchem
     Subscription = apps.get_model("zerver", "subscription")
     Recipient = apps.get_model("zerver", "recipient")
 
-    # Here we compute per-stream subscriber_count in the sub-query which returns a stream_subscribers_table,
-    # then Update it in-place using that computed value.
-    # The whole query is then executed by one batch per realm.
+    # Here we compute per-stream subscriber_count in the sub-query
+    # which returns a stream_subscribers_table, then Update it
+    # in-place using that computed value. The whole query is then
+    # executed by one batch per realm.
     query = sql.SQL(
         """UPDATE {stream_table} AS stream
-            SET subscriber_count = stream_subscribers_table.n_subscribers
+            SET subscriber_count = stream_subscribers_table.count
             FROM (
                 SELECT recipient.type_id AS stream_id,
-                COUNT(subscription.user_profile_id) AS n_subscribers
+                COUNT(subscription.user_profile_id) AS count
                 FROM {subscription_table} AS subscription
                 JOIN {recipient_table} AS recipient
                     ON subscription.recipient_id = recipient.id
```

timabbott commented on zerver/lib/stream_subscription.py:
```diff
@@ -313,3 +322,87 @@ def get_subscriptions_for_send_message(
+            SET subscriber_count = {stream_table}.subscriber_count + delta_table.delta
+            FROM (VALUES %s) AS delta_table(id, delta)
+            WHERE {stream_table}.id = delta_table.id;
+        """
```
I'm not entirely convinced that this algorithm is safe from races. But it may be good enough in practice. I think we can merge this and then investigate anomalies before actually making use of the value in production; after all, it's currently a field we don't read.

timabbott commented:
I did the cleanups noted above and am merging this, thanks for all the work on this @abdelrahman725!

You can get started on writing the code to put this value into the API.
